### PR TITLE
eSCO codec/HCI definitions + Host support 

### PR DESCRIPTION
**Pull request by zxzxwu**
I only added HCI packets and eSCO codec definitions, as well as host handler. For `device`, it's not easy to integrate SCO into current handlers.

I also tested the SCO HCI path on Android, with IBluetoothHci. Links can be created and controllers consistently report packets, but later packets are all 0. It's probably because the host didn't send any SCO packet?

Most USB dongles don't support eSCO, so I haven't tried to setup them legacy SCO.

### CODE REVIEW

**Comment by zxzxwu** [comment]
Some thoughts:
* Probably not all controllers support PCM IO? Also, may it reduce the bandwidth/quality?
* If we use transparent IO, ffplay/liblc3 may handle the co/decoding, but not for CVSD.

File: bumble/hfp.py [hunk 1]
```diff
@@ -747,3 +748,156 @@ def sdp_records(
...
+    SCO_CVSD_D1 = enum.auto()
+    ESCO_CVSD_S1 = enum.auto()
+    ESCO_CVSD_S2 = enum.auto()
+    ESCO_CVSD_S3 = enum.auto()
+    ESCO_CVSD_S4 = enum.auto()
+    ESCO_MSBC_T1 = enum.auto()
+    ESCO_MSBC_T2 = enum.auto()
+
+
+_ESCO_PARAMETERS_COMMON = {
```
Comment (barbibulle) [inline bumble/hfp.py:769]: maybe we could use a class (like a dataclass) instead of a dictionary here?

File: bumble/hci.py [hunk 2]
```diff
@@ -5653,6 +5716,64 @@ def __str__(self):
         )
 
 
+# -----------------------------------------------------------------------------
+class HCI_SynchronousDataPacket:
+    '''
+    See Bluetooth spec @ 5.4.2 HCI ACL Data Packets
```
Comment (barbibulle) [inline bumble/hci.py:5722]: Should be `5.4.3 HCI Synchronous Data packets`

File: bumble/hci.py [hunk 3]
```diff
@@ -5653,6 +5716,64 @@ def __str__(self):
...
+    '''
+
+    hci_packet_type = HCI_SYNCHRONOUS_DATA_PACKET
+
+    @staticmethod
+    def from_bytes(packet: bytes) -> 'HCI_SynchronousDataPacket':
+        # Read the header
+        h, data_total_length = struct.unpack_from('<HB', packet, 1)
+        connection_handle = h & 0xFFF
+        packat_status = (h >> 12) & 3
```
Comment (barbibulle) [inline bumble/hci.py:5732]: typo: packet

File: bumble/host.py [hunk 4]
```diff
@@ -484,6 +487,10 @@ def on_hci_acl_data_packet(self, packet):
         if connection := self.connections.get(packet.connection_handle):
             connection.on_hci_acl_data_packet(packet)
 
+    def on_hci_sco_data_packet(self, packet):
+        # Look for the connection to which this data belongs
```
Comment (barbibulle) [inline bumble/host.py:491]: I don't think this comment applies here.

**Comment by barbibulle** [comment]
> Some thoughts:
> 
> * Probably not all controllers support PCM IO? Also, may it reduce the bandwidth/quality?
> * If we use transparent IO, ffplay/liblc3 may handle the co/decoding, but not for CVSD.

We could implement CVSD in python, and leave other formats for external libraries (it would also be interesting to try and package liblc3 as a wasm binary and use it within python using a wasm interpreter)

**Comment by barbibulle** [comment]
> I only added HCI packets and eSCO codec definitions, as well as host handler. For `device`, it's not easy to integrate SCO into current handlers.

I have some ideas on how to support eSCO connections at the Device level. I can try to push up a PR for that once this one is finished and merged.

File: bumble/hfp.py (hunk 1 above)
Comment (zxzxwu) [reply bumble/hfp.py]: Modified

File: bumble/hci.py (hunk 2 above)
Comment (zxzxwu) [reply bumble/hci.py]: Done

File: bumble/hci.py (hunk 3 above)
Comment (zxzxwu) [reply bumble/hci.py]: Done

File: bumble/host.py (hunk 4 above)
Comment (zxzxwu) [reply bumble/host.py]: Done

**Review by uael — APPROVED** [review]

File: bumble/hfp.py [hunk 5]
```diff
@@ -747,3 +748,198 @@ def sdp_records(
...
+    input_transport_unit_size: int = 0
+    output_transport_unit_size: int = 0
+    input_bandwidth: int = 16000
+    output_bandwidth: int = 16000
+    transmit_bandwidth: int = 8000
+    receive_bandwidth: int = 8000
+    transmit_codec_frame_size: int = 60
+    receive_codec_frame_size: int = 60
+
+    def to_dict(self):
```
Comment (barbibulle) [inline bumble/hfp.py:812]: This is probably not necessary, since @dataclass already has `asdict` which does the same thing.

**Review by barbibulle — APPROVED** [review]

File: bumble/hfp.py (hunk 5 above)
Comment (zxzxwu) [reply bumble/hfp.py]: Done

**Pull request merged**